### PR TITLE
BUG: DataFrame reductions dtypes on object input

### LINKED ISSUE #49603: BUG: DataFrame reductions with object dtype and axis=1
In pandas 1.5.x and before, the dtype of the result is different when using `axis=1`:

```
df = pd.DataFrame({'a': [1, 2, 3], 'b': [4, 5, 6]}, dtype=object)
print(df.sum(axis=1, numeric_only=None))
print(df.sum(axis=1, numeric_only=False))
0    5.0
1    7.0
2    9.0
dtype: float64
0    5
1    7
2    9
dtype: object
```

When using `axis=0`, both the examples above result in object dtype. #49551 changed the `numeric_only=False` case to return `float64` in order to preserve the default behavior. However, it seems better to have the result be object dtype for both examples above instead.

### LINKED ISSUE #51242: BUG: `DataFrame.min(axis=1)` raises `FutureWarning` for timezone aware datetimes and returns wrong results
### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [ ] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [ ] I have confirmed this bug exists on the [main branch](https://pandas.pydata.org/docs/dev/getting_started/install.html#installing-the-development-version-of-pandas) of pandas.

### Reproducible Example

```python
import pandas as pd

s1 = pd.Series([pd.NaT, pd.NaT, datetime.datetime(1900, 1, 1)])
s2 = pd.Series([pd.NaT, datetime.datetime(1900, 1, 2), datetime.datetime(1900, 1, 3)])

df = pd.DataFrame({'c1':s1, 'c2':s2})
df_utc = pd.DataFrame({'c1':s1.dt.tz_localize("UTC"), 'c2':s2.dt.tz_localize("UTC")})

df.min(axis=1,skipna=False) # expected result with 2 NaT
df.min(axis=1,skipna=True) # expected result with 1 NaT

df_utc.min(axis=1,skipna=False # ISSUE unexpected result with 1 NaT instead of 2
df_utc.min(axis=1,skipna=True) # ISSUE warning + unexpected results: all NaNs 

```

### Issue Description

We would expect 2 NaTs in the following call, not just one
```
In [71]: df_utc.min(axis=1,skipna=False)
Out[71]: 
0                         NaT
1   1900-01-02 00:00:00+00:00
2   1900-01-01 00:00:00+00:00
dtype: datetime64[ns, UTC]
```

We would expect a single NaT, but we get a warning and a series of float NaNs
```
In [72]: df_utc.min(axis=1,skipna=True)
<ipython-input-72-c024a9c167c1>:1: FutureWarning: Dropping of nuisance columns in DataFrame reductions (with 'numeric_only=None') is deprecated; in a future version this will raise TypeError.  Select only valid columns before calling the reduction.
  df_utc.min(axis=1,skipna=True)
Out[72]: 
0   NaN
1   NaN
2   NaN
dtype: float64
```
The simliar calls with `df` (no timezones) return expected results.
```
In [73]: df.min(axis=1,skipna=False)
Out[73]: 
0          NaT
1          NaT
2   1900-01-01
dtype: datetime64[ns]

In [74]: df.min(axis=1,skipna=True)
Out[74]: 
0          NaT
1   1900-01-02
2   1900-01-01
dtype: datetime64[ns]
```

### Expected Behavior

see description

### Installed Versions

<details>

INSTALLED VERSIONS
------------------
commit           : 8dab54d6573f7186ff0c3b6364d5e4dd635ff3e7
python           : 3.8.10.final.0
python-bits      : 64
OS               : Linux
OS-release       : 5.4.191-el7.lime.2.x86_64
Version          : #1 SMP Thu Jul 14 19:09:52 EDT 2022
machine          : x86_64
processor        : x86_64
byteorder        : little
LC_ALL           : None
LANG             : en_US.UTF-8
LOCALE           : en_US.UTF-8

pandas           : 1.5.2
numpy            : 1.21.3
pytz             : 2022.1
dateutil         : 2.8.2
setuptools       : 65.2.0
pip              : 22.1.2
Cython           : None
pytest           : 4.6.5
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : 4.6.2
html5lib         : None
pymysql          : None
psycopg2         : 2.9.3
jinja2           : 3.0.3
IPython          : 7.33.0
pandas_datareader: None
bs4              : None
bottleneck       : None
brotli           : 
fastparquet      : None
fsspec           : None
gcsfs            : None
matplotlib       : 3.3.4
numba            : 0.50.1+3.g7f7b8af82
numexpr          : None
odfpy            : None
openpyxl         : 3.0.3
pandas_gbq       : None
pyarrow          : 4.0.1
pyreadstat       : None
pyxlsb           : None
s3fs             : None
scipy            : 1.7.3
snappy           : None
sqlalchemy       : 1.4.9
tables           : None
tabulate         : 0.8.9
xarray           : None
xlrd             : 2.0.1
xlwt             : 1.3.0
zstandard        : None
tzdata           : None

</details>

### PR DESCRIPTION
**Pull request by rhshadrach**
- [x] closes #49603 (Replace xxxx with the GitHub issue number)
- [x] closes #51242
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [x] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

File: doc/source/whatsnew/v2.0.0.rst [hunk 1]
```diff
@@ -1204,10 +1204,10 @@ Numeric
 ^^^^^^^
 - Bug in :meth:`DataFrame.add` cannot apply ufunc when inputs contain mixed DataFrame type and Series type (:issue:`39853`)
 - Bug in arithmetic operations on :class:`Series` not propagating mask when combining masked dtypes and numpy dtypes (:issue:`45810`, :issue:`42630`)
-- Bug in DataFrame reduction methods (e.g. :meth:`DataFrame.sum`) with object dtype, ``axis=1`` and ``numeric_only=False`` would not be coerced to float (:issue:`49551`)
 - Bug in :meth:`DataFrame.sem` and :meth:`Series.sem` where an erroneous ``TypeError`` would always raise when using data backed by an :class:`ArrowDtype` (:issue:`49759`)
 - Bug in :meth:`Series.__add__` casting to object for list and masked :class:`Series` (:issue:`22962`)
 - Bug in :meth:`DataFrame.query` with ``engine="numexpr"`` and column names are ``min`` or ``max`` would raise a ``TypeError`` (:issue:`50937`)
+- Bug in :meth:`DataFrame.min` and :meth:`DataFrame.max` with tz-aware data containing NA and ``axis=1`` would return incorrect results (:issue:`51242`)
```
Comment (jbrockmendel) [inline doc/source/whatsnew/v2.0.0.rst:1210]: does NA here refer to pd.NaT?

File: pandas/core/frame.py [hunk 2]
```diff
@@ -10461,54 +10460,44 @@ def _get_data() -> DataFrame:
...
-                    # try to coerce to the original dtypes item by item if we can
-                    pass
-
+        # Case with EAs see GH#35881
+        df = self
+        if numeric_only:
+            df = _get_data()
         if axis is None:
-            return result
+            return func(df.values)
```
Comment (jbrockmendel) [inline pandas/core/frame.py:10468]: hmm .values can be expensive, might be better to reduce twice?  (... which can also be expensive.  darn).  Is punting on this viable?

Comment (rhshadrach) [reply pandas/core/frame.py:10468]: Using `.values` is the current behavior on main if `numeric_only` is False and `axis` is not 0.

https://github.com/pandas-dev/pandas/blob/c7fa61113b4cf09581e8f31f6053f2e64b83a9fc/pandas/core/frame.py#L10494-L10496

In fact, main is currently broken when `numeric_only` is False and `axis` is None:

```
df = pd.DataFrame({'a': [1, 1, 2], 'b': [3, 4, 5], 'c': list('xyz')})
result = df.mean(axis=None, numeric_only=True)
print(result)
# a    1.333333
# b    4.000000
# dtype: float64

result2 = df[['a', 'b']].mean(axis=None)
print(result2)
# 2.6666666666666665
```

Assuming this is okay for now, I can add a test for when `numeric_only` is False and `axis` is None.

File: doc/source/whatsnew/v2.0.0.rst (hunk 1 above)
Comment (rhshadrach) [reply doc/source/whatsnew/v2.0.0.rst:1210]: Thanks - will fix.

File: pandas/core/frame.py (hunk 2 above)
Comment (rhshadrach) [reply pandas/core/frame.py:10468]: @jbrockmendel - friendly ping

File: pandas/core/internals/array_manager.py [hunk 3]
```diff
@@ -983,14 +983,10 @@ def reduce(self: T, func: Callable) -> T:
...
             # what if datetime results in timedelta? (eg std)
-            if res is NaT and is_timedelta64_ns_dtype(arr.dtype):
-                result_arrays.append(np.array(["NaT"], dtype="timedelta64[ns]"))
-            else:
-                # error: Argument 1 to "append" of "list" has incompatible type
-                # "ExtensionArray"; expected "ndarray"
-                result_arrays.append(
-                    sanitize_array([res], None)  # type: ignore[arg-type]
-                )
+            dtype = arr.dtype if res is NaT else None
```
Comment (jbrockmendel) [inline pandas/core/internals/array_manager.py:986]: havent looked too closely, but could this mess up on "std" with dt64?

File: pandas/core/nanops.py [hunk 4]
```diff
@@ -1533,7 +1533,15 @@ def _maybe_null_out(
                 result[null_mask] = None
     elif result is not NaT:
         if check_below_min_count(shape, mask, min_count):
-            result = np.nan
+            result_dtype = getattr(result, "dtype", None)
+            if is_float_dtype(result_dtype):
+                # Preserve dtype when possible
+                # mypy doesn't infer result_dtype is not None
+                result = getattr(
```
Comment (jbrockmendel) [inline pandas/core/nanops.py:1540]: `result_type.type("nan")`?

**Comment by jbrockmendel** [comment]
except for the ArrayManager thing, this LGTM

File: pandas/core/internals/array_manager.py (hunk 3 above)
Comment (rhshadrach) [reply pandas/core/internals/array_manager.py:986]: Is the following sufficient?

```
df = pd.DataFrame({'a': ['2022-01-01', '2022-01-02', pd.NaT, '2022-01-03']})
df['a'] = pd.to_datetime(df['a'])
print(df.std())
# a   1 days
# dtype: timedelta64[ns]
```

I don't think any tests hit _reduce with std and dt64. Can add.

Comment (jbrockmendel) [reply pandas/core/internals/array_manager.py:986]: i think the relevant test would have skipna=False (or all-NaT values)

Comment (rhshadrach) [reply pandas/core/internals/array_manager.py:986]: Thanks - test added.

Comment (jbrockmendel) [reply pandas/core/internals/array_manager.py:981]: test_std_datetime64_with_nat passes with the change you made here?  i may need more caffeine

Comment (rhshadrach) [reply pandas/core/internals/array_manager.py:981]: I see it now - thanks. I get the same behavior on main - that std should return a td64 but when the result is NaT and the input is dt64 we get dt64 instead. From what I can tell std i the only op that does dt64 -> td64 and the PR is doing no harm here. Okay to xfail?

Comment (jbrockmendel) [reply pandas/core/internals/array_manager.py:981]: Sure, thanks for taking a look

**Review by jbrockmendel — APPROVED** [review]
LGTM

**Review by mroeschke — APPROVED** [review]

**Comment by mroeschke** [comment]
Thanks @rhshadrach

**Comment by jbrockmendel** [comment]
@rhshadrach looks like this caused a regression in https://asv-runner.github.io/asv-collection/pandas/#stat_ops.FrameOps.time_op?p-op='mean'&p-dtype='Int64'&p-axis=1&commits=b836a88f for nullable cases with axis=1.  Can you take a look?

**Comment by rhshadrach** [comment]
1.5.3 is dropping down to the underlying NumPy array and calling the op with `axis=1`. main is computing the transpose of the DataFrame (which is inefficient for EAs), then computing the sum via the Block manager (which I _think_ is inefficient for EAs with wide data).

I'm still looking for a better alternatively currently, but likely I expect to need to largely restore the previous implementation and defer a better solution for 2.1.

**Comment by jbrockmendel** [comment]
Is it a correctness vs perf tradeoff?  If so correctness wins every time.

Might be able to do something similar to _reduce_axis1 to avoid .values/transpose

**Comment by rhshadrach** [comment]
> Is it a correctness vs perf tradeoff? If so correctness wins every time.

I think I can get every test to pass by using a combination of 1.5.x's `_reduce` and that in this PR except for test_minmax_tzaware_skipna_axis_1. For tzaware with NaT, dropping down to NumPy arrays is insufficient to compute the op. I don't see a way to fix this in any easy manner and avoid the perf regression. I think (but haven't confirmed yet) that this was also broken in 1.5.x; is it worth the perf hit for this?

Longer term (for 2.1), it does seem to me doing something like _reduce_axis1 is the way to go. For example, here is the ASV that identified this regression on 1.5.x (I'm leaving out the setup here) against a direct implementation of `mean` for axis=1:

```
%timeit df.mean(axis=1)
18.5 ms ± 40.4 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

%timeit sum(df[c] for c in df) / df.shape[1]
546 µs ± 55.6 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
```

**Comment by jbrockmendel** [comment]
Any interest in trying to get this improved for 2.0?

**Comment by jorisvandenbossche** [comment]
For the case of masked arrays (the specific case from the asv regression), the transpose spends most of the time in reconstructing an EA for each row, which spends a lot of time in `_coerce_to_data_and_mask`. There is a lot of room for improvement in that function, if we would want to do that (eg it's doing a lot of dtype checks, and we know that those are slow if you do those repeatedly many times, in some other (critical) paths we switched to `dtype.kind == ..` checks, IIRC)

Improving the constructor might be worth anyway, but I agree that for the specific case of the numeric reduction with axis=1, doing a special case like `_reduce_axis1` already does for any/all seems quite easy to do without being very complex. 
Because however we improve the constructor, constructing 10,000x a small EA in case of 10,000 rows DataFrame is just always going to be slower.

**Comment by rhshadrach** [comment]
> Might be able to do something similar to _reduce_axis1 to avoid .values/transpose

I looked into this - I think there are some common ops / dtypes we can do this for, but we can't do it for all ops (e.g. median) and all dtypes (e.g. 3rd party EAs since _I think_ the implementation will depend on knowing the value of the neutral element for an op like sum or prod).

A straightforward way to fix this regression is to mostly revert this PR, where we still return object dtype on object input but otherwise the results are unchanged. I can put up a PR for this approach if #51923 isn't the right way forward or is too much to try to get it into 2.0.

**Pull request merged**